app/src/org/commcare/gis/MapboxLocationPickerActivity.kt

```diff
@@ -0,0 +1,277 @@
+    private lateinit var locationController: CommCareLocationController
+    private val mapStyles = arrayOf(
+            Style.MAPBOX_STREETS,
+            Style.SATELLITE,
+            Style.SATELLITE_STREETS,
+            Style.OUTDOORS
+    )
+    private var currentMapStyleIndex = 0
+    private var cameraPosition: CameraPosition? = null
+    private var symbolManager: SymbolManager? = null
+    private var symbol: Symbol? = null
+    private val mapClickListener = MapboxMap.OnMapClickListener { point ->
+        isManualSelectedLocation = true
+        // Add marker.
+        updateMarker(point)
+        viewModel.reverseGeocode(point)
+        true
+    }
+
+    // don't reset marker to current GPS location if we manually selected a location
+    private var isManualSelectedLocation = false
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        viewModel = ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory.getInstance(application))
+                .get(MapboxLocationPickerViewModel::class.java)
+        attachListener()
+        locationController = CommCareLocationControllerFactory.getLocationController(this, this)
+
+        // Check extras
+        if (intent.hasExtra(GeoPointWidget.LOCATION)) {
+            val inputLocation = intent.getDoubleArrayExtra(GeoPointWidget.LOCATION)!!
+            cameraPosition = CameraPosition.Builder()
+                    .target(LatLng(inputLocation[0], inputLocation[1]))
+                    .zoom(10.0)
+                    .tilt(20.0)
+                    .build()
+        }
+        mapView.showCurrentLocationBtn(false)
+    }
+
+    private fun attachListener() {
+        confirm_location_button.visibility = if (inViewMode()) {
+            View.GONE
+        } else {
+            View.VISIBLE
+        }
+        confirm_location_button.setOnClickListener {
+            // Use the map target's coordinates to make a reverse geocoding search
+            Intent().apply {
+                putExtra(FormEntryConstants.LOCATION_RESULT, GeoUtils.locationToString(viewModel.getLocation()))
+            }.also {
+                setResult(RESULT_OK, it)
+                finish()
+            }
+        }
+        cancel_button.setOnClickListener {
+            finish()
+        }
+        switch_map.setOnClickListener {
+            currentMapStyleIndex = (currentMapStyleIndex + 1) % mapStyles.size
+            val style = Style.Builder()
+                    .fromUri(mapStyles[currentMapStyleIndex])
+                    .withImage(MARKER_ICON_IMAGE_ID, ContextCompat.getDrawable(this, R.drawable.marker)!!)
+            map.setStyle(style) {
+                val loc = viewModel.getLocation()
+                updateMarker(LatLng(loc.latitude, loc.longitude, loc.altitude))
+            }
+        }
+        current_location.setOnClickListener {
+            isManualSelectedLocation = false
```

Comment:
this line here should be sufficient , the location listener will take care of updating the location